Normalize stop times travel times for set of pattern stops

landonreed opened the pull request:
### Checklist

- [x] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [x] Any modified or new methods or classes have helpful JavaDoc and code is thoroughly commented
- [x] The description lists all applicable issues this PR seeks to resolve
- [x] The description lists any configuration setting(s) that differ from the default settings
- [x] All tests and CI builds passing

### Description

Adds feature to normalize all stop times for a set of pattern stops in order to simplify updating
existing trips in bulk when a pattern changes (e.g., travel times change or new pattern stops are
inserted).

evansiroky commented on src/test/java/com/conveyal/gtfs/loader/JDBCTableWriterTest.java:
```diff
@@ -484,6 +486,50 @@ public void canCreateUpdateAndDeleteFrequencyTripForFrequencyPattern() throws IO
+            new PatternStopDTO(patternId, lastStopId, 1)
+        };
+        patternStops[1].default_travel_time = initialTravelTime;
+        PatternDTO pattern = createRouteAndPattern(UUID.randomUUID().toString(), patternId, "Pattern A", null, new ShapePointDTO[]{}, patternStops, 0);
```
Please format this line so it isn't so long.

evansiroky commented on src/test/java/com/conveyal/gtfs/loader/JDBCTableWriterTest.java:
```diff
@@ -484,6 +486,50 @@ public void canCreateUpdateAndDeleteFrequencyTripForFrequencyPattern() throws IO
+        patternStops[1].default_travel_time = initialTravelTime;
+        PatternDTO pattern = createRouteAndPattern(UUID.randomUUID().toString(), patternId, "Pattern A", null, new ShapePointDTO[]{}, patternStops, 0);
+        // Create trip with travel times that match pattern stops.
+        TripDTO tripInput = constructTimetableTrip(pattern.pattern_id, pattern.route_id, startTime, initialTravelTime);
```
Please format this line so it isn't so long.

evansiroky replied:
actually, this line looks ok in the IDE, disregard.

evansiroky commented on src/test/java/com/conveyal/gtfs/loader/JDBCTableWriterTest.java:
```diff
@@ -484,6 +486,50 @@ public void canCreateUpdateAndDeleteFrequencyTripForFrequencyPattern() throws IO
+        JdbcTableWriter updateTripWriter = createTestTableWriter(tripsTable);
+        updateTripWriter.normalizeStopTimesForPattern(pattern.id, 0);
+        // Read pattern stops from database and check that the arrivals/departures have been updated.
+        JDBCTableReader<StopTime> stopTimesTable = new JDBCTableReader(Table.STOP_TIMES, testDataSource, testNamespace + ".", EntityPopulator.STOP_TIME);
```
Please format this line so it isn't so long.

evansiroky commented on src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java:
```diff
@@ -252,6 +254,34 @@ public String update(Integer id, String json, boolean autoCommit) throws SQLExce
+     */
+    public int normalizeStopTimesForPattern (int id, int beginWithSequence) throws SQLException {
+        try {
+            JDBCTableReader<PatternStop> patternStops = new JDBCTableReader(Table.PATTERN_STOP, dataSource, tablePrefix + ".", EntityPopulator.PATTERN_STOP);
```
Please format this line so it isn't so long.

evansiroky commented on src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java:
```diff
@@ -670,6 +702,73 @@ private int updateStopTimesForPatternStop(ObjectNode patternStop, int previousTr
+        if (serviceIdFilters != null && serviceIdFilters.size() > 0) {
+            getFirstTravelTimeSql += String.format(
+                " and t.service_id in (%s)",
+                Collections.nCopies(serviceIdFilters.size(), "?")
```
Does this work? Wouldn't this just create a bunch of question marks without commas? It doesn't appear that this filter is ever used anyways. Is it needed?

evansiroky replied:
Also, there aren't any prepared statement set statements for these extra parameters either.

evansiroky commented on src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java:
```diff
@@ -670,6 +702,73 @@ private int updateStopTimesForPatternStop(ObjectNode patternStop, int previousTr
+        PreparedStatement updateStopTimeStatement = connection.prepareStatement(updateTravelTimeSql);
+        LOG.info(updateStopTimeStatement.toString());
+        final BatchTracker stopTimesTracker = new BatchTracker("stop_times", updateStopTimeStatement);
+        for (String tripId : timesForTripIds.keySet()) {
```
Could this for loop be avoided by doing the update statement using a join to find all applicable trip IDs?

evansiroky replied:
Oh, hmm, I see why there are 2 loops. It could still probably be combined into one huge messy SQL query, but I suppose this is fine for now.

landonreed replied:
🤔... is `one huge messy SQL query` a desirable thing? I don't think it necessarily is.

evansiroky replied:
It's only desirable if there are considerable performance benefits. However, that would entail much more investigation in running jmeter tests and whatnot so yeah, I'm fine with keeping this as is until users ever start complaining about the performance of this.

abyrd replied:
And even if there are considerable performance benefits, it's only desirable if the code is factored to remain readable, and that performance gain is actually needed and a significant fraction of total runtime. Otherwise in my opinion the best thing to do is add a comment informing future readers of the alternative approach.

landonreed commented:
@evansiroky, I removed the service ID filter, which was really just a placeholder for a feature that I don't think we will implement. I also added another test that's not exactly related, but something that was missing and I wanted to verify.

landonreed commented:
:tada: This PR is included in version 4.3.0 :tada:

The release is available on [GitHub release](https://github.com/conveyal/gtfs-lib/releases/tag/v4.3.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
